whylog/log_reader/__init__.py

```diff
@@ -2,20 +2,68 @@
 
 import six
 
+from whylog.front import FrontInput
+from whylog.log_reader.exceptions import NoLogTypeError
+from whylog.log_reader.searchers import BacktrackSearcher
+
 
 @six.add_metaclass(ABCMeta)
 class AbstractLogReader(object):
     @abstractmethod
     def get_causes(self, front_input):
         pass
 
+    @abstractmethod
+    def get_causes_tree(self, front_input):
+        pass
+
 
 class LogReader(AbstractLogReader):
     def __init__(self, config):
         self.config = config
 
     def get_causes(self, front_input):
-        pass
+        input_log_type = self.config.get_log_type(front_input)
+        if not input_log_type:
+            raise NoLogTypeError(front_input)
+        investigation_plan = self.config.create_investigation_plan(front_input, input_log_type)
+        manager = SearchManager(investigation_plan)
+        return manager.investigate()
 
     def get_causes_tree(self, front_input):
         pass
+
+
+class SearchManager(object):
+    def __init__(self, investigation_plan):
+        self._investigation_plan = investigation_plan
+
+    def investigate(self):
+        clues = {}
+        for step, log_type in self._investigation_plan.get_next_investigation_step_with_log_type():
+            search_handler = SearchHandler(step, log_type)
+            clues.update(search_handler.investigate())
+        # TODO checking up the constraints should take place here
+        return [
+            FrontInput(
```

Comment:
Recently, we talked that this name is not fortunately. However I'm not sure that is best time for renaming.